### PR TITLE
WIP -- Entitlements for enterprise learners

### PR DESCRIPTION
**Pull request by saleem-latif**
@mattdrayer , @asadiqbal08 , @zubair-arbi  FYI

### CODE REVIEW

File: ecommerce/enterprise/helpers.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
+"""
+Helper methods for enterprise app.
```
Comment (mattdrayer) [inline ecommerce/enterprise/helpers.py:2]: We'll need to add an "is_feature_enabled" check at some point

File: ecommerce/enterprise/exceptions.py [hunk 2]
```diff
@@ -0,0 +1,9 @@
+"""
+Exception classes for enterprise app.
+"""
+
+
+class RequiresDataSharingError(Exception):
```
Comment (mattdrayer) [inline ecommerce/enterprise/exceptions.py]: RequiresDataSharingConsentError

File: ecommerce/settings/base.py [hunk 3]
```diff
@@ -599,3 +602,12 @@
...
 CRISPY_TEMPLATE_PACK = 'bootstrap3'
+
+
+# Enterprise settings
+# enable or disable enterprise features
+ENABLE_ENTERPRISES = True
+
+# name for waffle switch to use for disabling enterprise features on runtime.
+# Note: management command ignore this switch
+DISABLE_ENTERPRISES_ON_RUNTIME_SWITCH = "disable_theming_on_runtime"
```
Comment (mattdrayer) [inline ecommerce/settings/base.py]: looks like this needs to be fixed

File: ecommerce/extensions/basket/views.py [hunk 4]
```diff
@@ -74,6 +75,10 @@ def get(self, request):
                 msg = _('An error occurred while retrieving enrollment details. Please try again.')
                 return HttpResponseBadRequest(msg)
 
+        # If learner is eligible for entitlements, apply entitlements
+        if voucher is None and enterprise_helpers.is_learner_eligible_for_entitlements(request.user, request.site):
+            voucher = enterprise_helpers.get_entitlement_voucher(request, product)
```
Comment (mattdrayer) [inline ecommerce/extensions/basket/views.py:80]: maybe we can move the eligibility check inside of get_entitlement_voucher?

**Comment by mattdrayer** [comment]
Overall this prototype makes sense to me, although I see two possible approaches:

1. Introduce the enterprise concept into each service where pertinent workflow orchestration is required.  Keep the orchestration logic contained inside of the enterprise app.  Minimize dependencies and completely refuse/reject direct coupling to application constructs.

2) Keep ALL enterprise orchestration logic within the enterprise app currently hosted in the LMS.  Execute workflows in other services via API calls from the enterprise app to each service.

I think we may want to go the latter route here, versus adding a "slice" of enterprise orchestration to each service where it is needed.  Thoughts?

**Comment by saleem-latif** [comment]
@mattdrayer If we are not doing a lot of customizations (branding wise of behavior wise) in ecommerce then we can go with the second approach. However, for customizing ecommerce, e.g. customized basket page for enterprise learners or having learners prompted for data sharing consent in ecommerce, we will have to add enterprise concept there.

Second approach, however, does seem a better way to go and I think for current use cases will be a good way forward.

**Comment by clintonb** [comment]
Please delete your branches after closing PRs.

**Pull request closed without merging**